docs/proposals/TLSPolicy.md

```diff
@@ -0,0 +1,84 @@
+# WIP
+# TLSPolicy API
+Currently, TLS certificates issued by the MGC are issued and managed using `cert-manager`. Configuration for these certificates are opaque to the end-user - we'd like to gradually expose a TLSPolicy to cover a range of use cases where administrators require more flexible configuration.
+
+## Problem
+
+Gateway Administrators require a comprehensive way to manage Automated, ACME-based TLS for Gateways. As gateways multiply and scale, there's a need for a flexible, policy-based API for handling Automatic TLS. Currently (July 2023), the MGC sets up automated TLS whenever a Gateway listener is created. There are currently no hooks into the creation of automated TLS certificates, nor user-exposed configuration options in terms ACME providers, custom renewal windows/schedules, key size, algorithm and other configuration options available via `cert-manager`.
+
+## Use Cases
+
+**Choosing a TLS Provider**: As an admin, I want to select a specific ACME provider for the TLS certificates in a particular gateway.
+
+**Setting the TLS Certificate Renewal Window**: As an admin, I need to define the renewal period for TLS certificates (e.g., every 90 days) and a lead time for renewals (e.g., 15 days before expiration).
+
+**Choosing a TLS Strategy**: As an admin, I want to employ different strategies for managing certificates (e.g., multiple host certificate or per host certificate).
+
+**Choosing a TLS Strategy**: As an admin, I want to be able to apply enterprise security standards to certificates generated by the MGC. I want to be able to specify certificate key size, algorithm and encoding.
+
+
+## Current implementation
+
+
+
+## Goals
+
+- Design an API that allows admins to handle Automatic TLS for any given gateway.
+- Facilitate easy selection of a TLS provider.
+- Allow admins to set the renewal window for TLS certificates.
+- Enable the use of different TLS strategies.
+- Consider additional TLS configurations such as enforcing HTTPS listeners and supporting wildcard hosts.
+
+## Non-Goals
+
+- Detailed implementation of the API.
+- Developing a working prototype of the API.
+- Identifying specific technologies or tools for implementing the API.
+- Abstracting cert-manager Issuer/ClusterIssuer configuration.
```

Comment:
On this, is this in reality going to be a CertManagerTLSPolicy? Regardless of naming, i expect it is, so would we ever want to abstract these?

Reply:
This is in non goals

Reply:
Yeah, i suppose since it was here it felt like it was something you aren't doing now but might do later.

If this is tightly coupled to cert-manager, I'm wondering do we need to reflect that in the naming of it, or at least make it more obvious in the wording of this proposal. Probably not, prefer just TLSPolicy tbh, but just putting it out there.